From c8eee499efa5755f252ba2b55865c23390231ba0 Mon Sep 17 00:00:00 2001
From: Vladimir Oltean <vladimir.oltean@nxp.com>
Date: Tue, 28 Jan 2020 23:55:17 +0200
Subject: [PATCH 1/1] arm64: dts: fsl: ls1028a-rdb: enable swp5 and move NPI
 port to it

This patch moves the NPI port from swp4 to swp5 on the Felix switch, and
makes the 2.5 port pair (eno2/swp4) operate in plain untagged mode
(previously it worked in DSA tagging mode).

The switch engine for 802.1CB FRER uses the MAC table (FDB) to assign
packets to a Seamless Stream ID (SSID). In turn, the MAC table needs the
Analyzer module (ANA) to inspect the frames.

But traffic that passes to/from the NPI port (which is what the DSA
driver uses as a CPU port) necessarily has the BYPASS flag set in the
injection header, which means "bypass the analyzer module".

So traffic originated from the NPI port will not be correctly assigned
to a SSID for 802.1CB.

As a workaround, we enable the other port pair and use that for 802.1CB
traffic that needs to be terminated on the local CPU.

Before:                               After:

    eno2    eno3                         eno2      eno3
     |       |                   Untagged |         |
     | DSA   x disabled           traffic |         | DSA
     | tags  |                  including |         | tags
     |       |                    802.1CB |         |
    swp4    swp5                         swp4      swp5

Of course, in this context "untagged" means "not tagged with an
injection/extraction DSA tag", it has nothing to do with the VLAN or
802.1CB R-Tag.

Signed-off-by: Vladimir Oltean <vladimir.oltean@nxp.com>
---
 .../boot/dts/freescale/fsl-ls1028a-rdb.dts    | 23 +++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts b/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts
index 2c92bcdb0d62..d396457e9a00 100644
--- a/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts
@@ -239,6 +239,16 @@
 	status = "disabled";
 };
 
+&enetc_port2 {
+	fixed-link {
+		pause;
+	};
+};
+
+&enetc_port3 {
+	status = "okay";
+};
+
 &enetc_mdio_pf3 {
 	qsgmii_phy1: ethernet-phy@10 {
 		reg = <0x10>;
@@ -292,6 +302,19 @@
 	};
 };
 
+&mscc_felix_port4 {
+	/delete-property/ ethernet;
+
+	fixed-link {
+		pause;
+	};
+};
+
+&mscc_felix_port5 {
+	status = "okay";
+	ethernet = <&enetc_port3>;
+};
+
 &sai4 {
 	status = "okay";
 };
-- 
2.25.1

